sympy/functions/special/bessel.py

```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         _, e = arg.as_coeff_exponent(x)
 
         if e.is_positive:
-            term_one = ((-1)**(nu -1)*log(z/2)*besseli(nu, z))
-            term_two = (z/2)**(-nu)*factorial(nu - 1)/2 if (nu).is_positive else S.Zero
-            term_three = (-1)**nu*(z/2)**nu/(2*factorial(nu))*(digamma(nu + 1) - S.EulerGamma)
-            arg = Add(*[term_one, term_two, term_three]).as_leading_term(x, logx=logx)
-            return arg
+            if nu is S.Zero:
+                # Equation 9.6.8 of Abramowitz and Stegun (10th ed, 1972).
+                term = -log(z) - S.EulerGamma + log(2)
+            else:
+                # Equation 9.6.9 of Abramowitz and Stegun (10th ed, 1972).
+                term = gamma(Abs(nu))*(z/2)**(-Abs(nu))/2
```

Comment:
This is checking if `nu is S.Zero` but what if `nu` is an expression that is possibly zero?

Is it possible to handle that case or do we need to know definitely whether `nu` is zero or not?

Reply:
I think we need to know before,  if `nu` is `zero` before proceeding or not here, so I think this works.
And also not sure if we were looking for catering those cases where `nu` might be an expression here.

So basically my PR (https://github.com/sympy/sympy/pull/23844) implemented basic series for bessel functions for example `bessely(1, sin(x))` but I forgot to consider the important case that `nu` might be non-integer which this PR is fixing.

Reply:
So what happens if `nu` is just a symbol?

Does it just return a series that will be valid assuming that the symbol is not zero?

Neglecting a degenerate case like the symbol being zero does not seem completely unreasonable to me but we need to clear if that is the expected behaviour or not.

Lower down there is:
```
if nu.is_integer:
   # integer case
else:
   # non-integer case
```
This is more problematic. For a symbol `nu.is_integer` will likely return None and this will take the non-integer case.

Do we want to use the non-integer case whenever `nu` is a symbol?

You might notice around the codebase things like this to handle three-valued logic:
```python
if nu.is_integer:
   # integer case
elif nu.is_integer is False:
   # non-integer case
else:
   # Here we do not know if it is an integer or not.
   # We need to decide how to handle this somehow.
   # Ideally this case should not return likely incorrect results.
   # Perhaps we could raise an error or perhaps there is
   # some other routine that we can fall back on or perhaps
   # the result could be unevaluated or something.
```

Reply:
Ahh yes, we should be tackling this. Just wasn't sure if we should do it through this pr or not.

Yeah I think we return a general series for besselk(nu, m) not involving any checks for zero etc

Reply:
I have now modified the `nu is S.Zero` block to raise a `NotImplementedError` if we do not know whether the symbol is zero or nonzero. Similarly, I have done
```
if nu.is_integer:
   # integer case
elif nu.is_noninteger:
   # non-integer case
#fallback appears later on
```

Series expansion currently (even on master) raises a `NotImplementedError` error if the order of the bessel function is symbolic (see https://github.com/sympy/sympy/pull/25664#issuecomment-1732226476).

I am not sure about how to correctly implement series expansion for symbolic order, given that for Bessel functions, the number of terms in the series depends on the order (so it does not seem it can be handled by simply returning a `Piecewise` instance for the integer and non-integer cases).

Reply:
> I am not sure about how to correctly implement series expansion for symbolic order, given that for Bessel functions, the number of terms in the series depends on the order (so it does not seem it can be handled by simply returning a `Piecewise` instance for the integer and non-integer cases).

Okay well let's make sure at least that we error in the correct way rather than returning incorrect results then.

Although can it not be a Piecewise series like:
```python
In [1]: Piecewise((exp(x).series(n=3), x>0), (sin(x).series(n=3), True))
Out[1]: 
⎧         2                   
⎪        x     ⎛ 3⎞           
⎪1 + x + ── + O⎝x ⎠  for x > 0
⎨        2                    
⎪                             
⎪         ⎛ 3⎞                
⎩    x + O⎝x ⎠       otherwise
```

Reply:
I am now trying to modify `besselk._eval_nseries` to return a Piecewise, which I believe forces me to implement series expansion for all possible cases (I can't raise `NotImplementedError`). I have a question about the expected output for series like
![equation 9.7.2 of Abramowitz and Stegun](https://github.com/user-attachments/assets/61f5cadb-58e1-4259-bd73-8f06c0f294e5)
(this is for expansion about z=infinity, which would correspond to the currently unimplemented `exp < 0` branch of `besselk._eval_nseries`). Is it okay to return unexpanded prefactors like the `exp(-z)` in the equation above? I think this is what users would expect, but are there any assumptions made elsewhere in the code (e.g. that the output is a polynomial in x) that this would break?

Note that
```
In [100]: ( exp(-x) ).series(x, oo, n=4)
Out[100]: 
 ⎛1        ⎞
O⎜──; x → ∞⎟
 ⎜ 4       ⎟
 ⎝x        ⎠
```

Reply:
It might not be a good idea to use a Piecewise here. As @anutosh491 said above:

> Ahh yes, we should be tackling this. Just wasn't sure if we should do it through this pr or not.

I'm just musing about the possibilities. For now let's just make sure that the conditions are being checked properly before returning the series in each case.

Reply:
yes @Kishore96in, we might not want to address everything through this PR. Just have branches for all cases and we can return "NotImplemented" there. I do like most of the work here and I did approve the standard non-integer cases a while back too !

Reply:
> > I am not sure about how to correctly implement series expansion for symbolic order, given that for Bessel functions, the number of terms in the series depends on the order (so it does not seem it can be handled by simply returning a `Piecewise` instance for the integer and non-integer cases).
> 
> Okay well let's make sure at least that we error in the correct way rather than returning incorrect results then.
> 

With my latest changes,
```
In [8]: besselk(y,x).series(x)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[8], line 1
----> 1 besselk(y,x).series(x)

File /usr/lib/python3.12/site-packages/sympy/core/expr.py:3020, in Expr.series(self, x, x0, n, dir, logx, cdir)
   3017 if x.is_positive is x.is_negative is None or x.is_Symbol is not True:
   3018     # replace x with an x that has a positive assumption
   3019     xpos = Dummy('x', positive=True)
-> 3020     rv = self.subs(x, xpos).series(xpos, x0, n, dir, logx=logx, cdir=cdir)
   3021     if n is None:
   3022         return (s.subs(xpos, x) for s in rv)

File /usr/lib/python3.12/site-packages/sympy/core/expr.py:3028, in Expr.series(self, x, x0, n, dir, logx, cdir)
   3026 from sympy.series.order import Order
   3027 if n is not None:  # nseries handling
-> 3028     s1 = self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
   3029     o = s1.getO() or S.Zero
   3030     if o:
   3031         # make sure the requested order is returned

File /usr/lib/python3.12/site-packages/sympy/functions/special/bessel.py:785, in besselk._eval_nseries(self, x, n, logx, cdir)
    783         return Add(*a) + Add(*b) + o
    784     else:
--> 785         raise NotImplementedError("besselk expansion is only implemented for real order")
    787 return super(besselk, self)._eval_nseries(x, n, logx, cdir)

NotImplementedError: besselk expansion is only implemented for real order
```
Is the error message okay?